### PR TITLE
Allow reordering of filter conditions for better performance

### PR DESCRIPTION
**Pull request by nickitat**
### Changelog category (leave one):
- Backward Incompatible Change

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Reordering of filter conditions from `[PRE]WHERE` clause is now allowed by default. It could be disabled by setting `allow_reorder_prewhere_conditions` to `false`.

---

For some time we already had it. But then some users reported that they started to see queries failing with exceptions after upgrade. Examples of such queries could be found in [this PR](https://github.com/ClickHouse/ClickHouse/pull/69560) that addressed the problem and restored old behaviour. ATST reordering might be quite beneficial for performance, especially when reading takes significant portion of query execution time.
So the idea of this PR is to enable reordering by default, but also introduce a setting that will allow to disable it in case some queries were dependent on the order of filters.

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 483c2677a023b00579edc9abb36d827e5cad9819 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/70657/483c2677a023b00579edc9abb36d827e5cad9819/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/ad7c74c549343391b5320bcf95dc2b44b4f04d40/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Core/SettingsChangesHistory.cpp [hunk 1]
```diff
@@ -100,6 +100,7 @@ static std::initializer_list<std::pair<ClickHouseVersion, SettingsChangesHistory
             {"distributed_cache_data_packet_ack_window", 5, 5, "A setting for ClickHouse Cloud"},
             {"allow_experimental_refreshable_materialized_view", false, true, "Not experimental anymore"},
             {"max_parts_to_move", 1000, 1000, "New setting"},
+            {"allow_reorder_prewhere_conditions", true, true, "New setting"},
```
Comment (Algunenano) [inline src/Core/SettingsChangesHistory.cpp:103]: This should maintain compatibility and be `false, true`, right?

Comment (nickitat) [reply src/Core/SettingsChangesHistory.cpp:103]: sure, my bad

**Review by Algunenano — APPROVED** [review]

**Comment by rschu1ze** [comment]
@nickitat For my understanding:
- To find the new filter order, what heuristics do we use? I guess the most selective predicate should be evaluated first but how do we know exactly which one is the most selective one?
- Reordering was previously disabled because queries failed. This sounds funny, (in theory) reordering should not have an effect on correctness at all. The only scenario I can imagine is caused by short-circuiting: `alwaysEvaluatesToFalse(x) AND alwaysThrows(y)` vs. `alwaysThrows(y) AND alwaysEvaluatesToFalse(x)` will return different results.
- Not sure if you know, but does re-ordering only happen in PREWHERE or also for multiple filters on top of joined tables (e.g. for filters that could not be pushed down)?

**Pull request merged**